airflow/utils/helpers.py

```diff
@@ -310,3 +310,53 @@ def exactly_one(*args) -> bool:
             "Not supported for iterable args. Use `*` to unpack your iterable in the function call."
         )
     return sum(map(bool, args)) == 1
+
+
+def prune_dict(val: Any, mode='strict'):
+    """
+    Given dict ``val``, returns new dict based on ``val`` with all
+    empty elements removed.
+
+    What constitutes "empty" is controlled by the ``mode`` parameter.  If mode is 'strict'
+    then only ``None`` elements will be removed.  If mode is ``truthy``, then element ``x``
+    will be removed if ``bool(x) is False``.
+    """
+    if mode == 'truthy':
+
+        def is_empty(x):
+            return bool(x) is False
+
+    elif mode == 'strict':
+
+        def is_empty(x):
+            return x is None
+
+    else:
+        raise ValueError("allowable values for `mode` include 'truthy' and 'strict'")
+
```

Comment:
Instead of multiple function definitions can we do the following instead, WDYT? 

```suggestion
def is_empty(x):
    if mode == 'truthy':
        return bool(x) is False
    if mode == 'strict':
        return x is None
    raise ValueError("allowable values for `mode` include 'truthy' and 'strict'")
```

Reply:
or the following and pass `mode` explicitly. Either way should be fine as the function definition is local to `prune_dict`:

```suggestion
def is_empty(x, mode='truthy'):
    if mode == 'truthy':
        return bool(x) is False
    if mode == 'strict':
        return x is None
    raise ValueError("allowable values for `mode` include 'truthy' and 'strict'")
```

Reply:
yeah the reason i went this way is cus the mode of `is_empty` will never change within a call of prune_dict.  and so i thought it would be an optimization to define once and not have to evaluate the mode in every call to is_empty

BUT, i realize now that in each recursion step we have to define again anyway.  we could make it so that we only define once  _and_ we never have to reevaluate by moving the actual work to another inner function, BUT, this results in essentially no performance gain anyway :)